node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainerTest.java

```diff
@@ -0,0 +1,327 @@
+            maintainer = new SpareCapacityMaintainer(deployer, nodeRepository, metric, Duration.ofDays(1), maxIterations);
+        }
+
+        private void addHosts(int count, NodeResources resources) {
+            List<Node> hosts = new ArrayList<>();
+            for (int i = 0; i < count; i++) {
+                Node host = nodeRepository.createNode("host" + hostIndex,
+                                                      "host" + hostIndex + ".yahoo.com",
+                                                      ipConfig(hostIndex + nodeIndex, true),
+                                                      Optional.empty(),
+                                                      new Flavor(resources),
+                                                      Optional.empty(),
+                                                      NodeType.host);
+                hosts.add(host);
+                hostIndex++;
+            }
+            hosts = nodeRepository.addNodes(hosts, Agent.system);
+            hosts = nodeRepository.setReady(hosts, Agent.system, "Test");
+            var transaction = new NestedTransaction();
+            nodeRepository.activate(hosts, transaction);
+            transaction.commit();
+        }
+
+        private void addNodes(int id, int count, NodeResources resources, int hostOffset) {
+            List<Node> nodes = new ArrayList<>();
+            ApplicationId application = ApplicationId.from("tenant" + id, "application" + id, "default");
+            for (int i = 0; i < count; i++) {
+                ClusterMembership membership = ClusterMembership.from(ClusterSpec.specification(ClusterSpec.Type.content, ClusterSpec.Id.from("cluster" + id))
+                                                                                 .group(ClusterSpec.Group.from(0))
+                                                                                 .vespaVersion("7")
+                                                                                 .build(),
+                                                                      i);
+                Node node = nodeRepository.createNode("node" + nodeIndex,
+                                                      "node" + nodeIndex + ".yahoo.com",
+                                                      ipConfig(hostIndex + nodeIndex, false),
+                                                      Optional.of("host" + ( hostOffset + i) + ".yahoo.com"),
+                                                      new Flavor(resources),
+                                                      Optional.empty(),
+                                                      NodeType.tenant);
+                node = node.allocate(application, membership, node.resources(), Instant.now());
+                nodes.add(node);
+                nodeIndex++;
+            }
+            nodes = nodeRepository.addNodes(nodes, Agent.system);
+            for (int i = 0; i < count; i++) {
```

Comment:
Why is this needed?

Reply:
The second loop? This whole thing is in lieu of allocating properly, because here I want full control over where nodes are placed, without being dependent on constraints applied during allocation. The second loop is needed because setting nodes ready will remove the allocation if I set it at creation.